### PR TITLE
[NEW] Add/Create/Remove channel on a team

### PR DESCRIPTION
**Pull request by gerzonc**
## Proposed changes

Adds the required functionality to add channels to a team, create channels and add them to a team (directly) and remove the channels from a team.

- Add channel to team let's you add more than one channel to a team at once
- Create channel let's you create a channel and directly add it to a team
- Remove channel from team let's you remove a channel from the team and keep it from outside

Depends on #3082 
## Screenshots
<table>
<thead>
<tr>
<th>Add channel to Team</th>
<th>Add Existing Channels to Team</th>
<th>Auto-join/Remove/Delete Channel</th>
<th>Removing Channel from Team</th>
<th>Deleting Channel from Team</th>
</thead>
</tr>
<tbody>
<tr>
<td>
image
</td>
<td>
image
</td>
<td>
image
</td>
<td>
image
</td>
<td>
image
</td>
</tr>
</tbody>
</table >

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/utils/touch.js [hunk 1]
```diff
@@ -15,29 +15,32 @@ class Touch extends React.Component {
 
 	render() {
 		const {
-			children, onPress, theme, underlayColor, ...props
+			children, onPress, onLongPress, theme, style, ...props
 		} = this.props;
 
 		return (
-			<RectButton
+			<Pressable
```
Comment (diegolmello) [inline app/utils/touch.js:22]: Why is this necessary?
That's a serious refactor and I'm not sure Pressable is that stable already.

File: app/views/TeamChannelsView.js [hunk 2]
```diff
@@ -201,14 +221,11 @@ class TeamChannelsView extends React.Component {
 		options.headerRight = () => (
 			<HeaderButton.Container>
 				<HeaderButton.Item iconName='search' onPress={this.onSearchPress} />
+				{ showCreate
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:224]: On Figma, this button is on left.

File: app/sagas/createChannel.js [hunk 3]
```diff
@@ -49,14 +71,22 @@ const handleRequest = function* handleRequest({ data }) {
 				encrypted
 			});
 			sub = yield call(createChannel, data);
-		}
 
+			if (data.teamId) {
```
Comment (diegolmello) [inline app/sagas/createChannel.js:75]: Front end started using `groups.create` and `channels.create` for this.
They added a new param to the api to send a `teamId`, so this extra step isn't needed.
image

File: app/utils/goRoom.js [hunk 4]
```diff
@@ -8,17 +8,36 @@ const navigate = ({ item, isMasterDetail, ...props }) => {
...
-		name: RocketChat.getRoomTitle(item),
-		t: item.t,
-		prid: item.prid,
-		room: item,
-		search: item.search,
-		visitor: item.visitor,
-		roomUserId: RocketChat.getUidDirectMessage(item),
-		...props
-	});
+	if (item.isTeamChannel) {
```
Comment (diegolmello) [inline app/utils/goRoom.js:11]: This is changing an abstraction that makes navigating to room to work both on mobile and tablet.
If you check other places, those extra steps are done before calling `goRoom`.

File: app/views/AddChannelTeamView.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	buttonText: {
+		fontSize: 17,
+		...sharedStyles.textRegular
+	},
+	buttonContainer: {
+		paddingVertical: 25
+	}
+});
+
+class AddChannelTeamView extends React.Component {
```
Comment (diegolmello) [inline app/views/AddChannelTeamView.js:36]: It's up to you, but WDYT about writing this in hooks instead of class?

File: app/views/AddExistingChannelView.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+			this.setState({ loading: false });
+		}
+	}
+
+	renderChannel = ({
+		onPress, testID, title, icon, checked
+	}) => {
+		const { theme } = this.props;
+		return (
+			<Touch
```
Comment (diegolmello) [inline app/views/AddExistingChannelView.js:173]: Can't you use `List.Item`?

**Review by diegolmello — CHANGES_REQUESTED** [review]
- [x] "Create new" didn't navigate to the channel
image

File: app/views/AddExistingChannelView.js [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+		const { selected } = this.state;
+		return selected.includes(rid);
+	}
+
+	toggleChannel = (rid) => {
+		const { selected } = this.state;
+
+		animateNextTransition();
+		if (!this.isChecked(rid)) {
+			// logEvent(events.SELECTED_USERS_ADD_USER);
```
Comment (diegolmello) [inline app/views/AddExistingChannelView.js:176]: ?

File: app/views/TeamChannelsView.js [hunk 8]
```diff
@@ -287,6 +304,93 @@ class TeamChannelsView extends React.Component {
 		}
 	}, 1000, true);
 
+	options = item => ([
+		{
+			title: I18n.t('Auto-join'),
+			icon: item.t === 'p' ? 'channel-private' : 'channel-public'
+			// onPress: this.autoJoin
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:311]: ?

File: app/presentation/RoomItem/Chip.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
```
Comment (diegolmello) [inline app/presentation/RoomItem/Chip.js:1]: Chips aren't applied
image
image

File: app/presentation/RoomItem/Chip.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import { Text, View } from 'react-native';
+import PropTypes from 'prop-types';
+
+import styles from './styles';
+import { themes } from '../../constants/colors';
+
+const Chip = React.memo(({ name, theme }) => (
```
Comment (diegolmello) [inline app/presentation/RoomItem/Chip.js:8]: Missing stories. Remember to test the overflow.

Comment (diegolmello) [inline app/presentation/RoomItem/Chip.js:8]: Rename it to `Tag` as there's no visual `delete` on the component

File: app/presentation/RoomItem/Chip.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+	<View style={{ backgroundColor: themes[theme].auxiliaryText, padding: 8 }}>
```
Comment (diegolmello) [inline app/presentation/RoomItem/Chip.js:9]: Remove the inline style for padding

File: app/presentation/RoomItem/Chip.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+		<Text
+			style={[
+				styles.title,
```
Comment (diegolmello) [inline app/presentation/RoomItem/Chip.js:12]: It's not the same style as RoomItem.Title.
image

File: app/presentation/RoomItem/RoomItem.js [hunk 13]
```diff
@@ -88,6 +92,9 @@ const RoomItem = ({
 								hideUnreadStatus={hideUnreadStatus}
 								alert={alert}
 							/>
+							{
+								autoJoin ? <Chip name='Auto-Join' /> : null
```
Comment (diegolmello) [inline app/presentation/RoomItem/RoomItem.js:96]: Missing I18n.

File: app/presentation/RoomItem/Touchable.js [hunk 14]
```diff
@@ -203,54 +217,71 @@ class Touchable extends React.Component {
...
 
 			return (
-
-				<PanGestureHandler
-					minDeltaX={20}
-					onGestureEvent={this._onGestureEvent}
-					onHandlerStateChange={this._onHandlerStateChange}
-					enabled={swipeEnabled}
-				>
+				<LongPressGestureHandler onHandlerStateChange={(this.onLongPressHandlerStateChange)}>
```
Comment (diegolmello) [inline app/presentation/RoomItem/Touchable.js:239]: I didn't get the extra parenthesis.

File: app/sagas/createChannel.js [hunk 15]
```diff
@@ -67,14 +71,17 @@ const handleRequest = function* handleRequest({ data }) {
 				encrypted
 			});
 			sub = yield call(createChannel, data);
+			if (sub.teamId) {
```
Comment (diegolmello) [inline app/sagas/createChannel.js:74]: Why is this inside `channel` creating logic instead of `team`'s?

File: app/sagas/createChannel.js [hunk 16]
```diff
@@ -104,7 +111,7 @@ const handleSuccess = function* handleSuccess({ data }) {
 	if (isMasterDetail) {
 		Navigation.navigate('DrawerNavigator');
 	}
-	goRoom({ item: data, isMasterDetail });
+	goRoom({ item: data.success ? data.team : data, isMasterDetail });
```
Comment (diegolmello) [inline app/sagas/createChannel.js:114]: Bad resolved conflicts

File: app/utils/goRoom.js [hunk 17]
```diff
@@ -9,9 +9,9 @@ const navigate = ({ item, isMasterDetail, ...props }) => {
 	}
 
 	navigationMethod('RoomView', {
-		rid: item.rid,
+		rid: item.roomId || item.rid,
```
Comment (diegolmello) [inline app/utils/goRoom.js:12]: Issue on resolve conflicts

File: app/views/TeamChannelsView.js [hunk 18]
```diff
@@ -287,13 +307,132 @@ class TeamChannelsView extends React.Component {
 		}
 	}, 1000, true);
 
+	options = item => ([
+		{
+			title: I18n.t('Auto-join'),
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:312]: It's not showing the checkbox on the right.

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/sagas/createChannel.js (hunk 15 above)
Comment (gerzonc) [reply app/sagas/createChannel.js:74]: Because `groups.create` and `channels.create` now can return a `teamId`, I'm using this validation here so *if* we receive a `teamId` then we will then add the room to the team with `addRoomsToTeam`

File: app/presentation/RoomItem/RoomItem.js [hunk 19]
```diff
@@ -88,6 +92,9 @@ const RoomItem = ({
 								hideUnreadStatus={hideUnreadStatus}
 								alert={alert}
 							/>
+							{
+								autoJoin ? <Tag name='Auto-Join' /> : null
```
Comment (diegolmello) [inline app/presentation/RoomItem/RoomItem.js:96]: I18n

**Review by diegolmello — CHANGES_REQUESTED** [review]

**Comment by diegolmello** [comment]
@gerzonc create channel isn't navigating to the correct RoomView

File: app/views/AddExistingChannelView.js [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+
+	// eslint-disable-next-line react/sort-comp
+	init = async() => {
+		try {
+			const { addTeamChannelPermission } = this.props;
+			const db = database.active;
+			const channels = await db.collections
+				.get('subscriptions')
+				.query(
+					Q.and(Q.where('team_id', ''), Q.or(Q.where('t', 'c'), Q.where('t', 'p'))),
```
[the remark this answers is not in the document]
Comment (gerzonc) [reply app/views/AddExistingChannelView.js:86]: Due to `team_id` being an indexed field its default value is set to `' '`

File: app/views/TeamChannelsView.js [hunk 21]
```diff
@@ -287,6 +290,107 @@ class TeamChannelsView extends React.Component {
...
+		},
+		{
+			title: I18n.t('Delete'),
+			icon: 'delete',
+			danger: true,
+			onPress: () => this.delete(item)
+		}
+	])
+
+	autoJoin = async(item) => {
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:314]: `toggleAutoJoin`

File: app/views/TeamChannelsView.js [hunk 22]
```diff
@@ -287,6 +290,107 @@ class TeamChannelsView extends React.Component {
...
+
+	removeRoom = async(item) => {
+		try {
+			const { data } = this.state;
+			const result = await RocketChat.removeTeamRoom({ roomId: item._id, teamId: this.team.teamId });
+			if (result.success) {
+				const newData = data.filter(room => result.room._id !== room._id);
+				this.setState({ loading: true, data: newData }, () => {
+					this.load();
+					this.loadTeam();
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:359]: Just filter from the state and this is not going to be needed

File: storybook/stories/Tag.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable import/no-extraneous-dependencies, import/no-unresolved, import/extensions, react/prop-types */
+import React from 'react';
```
Comment (diegolmello) [inline storybook/stories/Tag.js:2]: Move to RoomItem stories

File: storybook/stories/Tag.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+import { View } from 'react-native';
+import { storiesOf } from '@storybook/react-native';
+
+import Tag from '../../app/presentation/RoomItem/Tag';
+
+const stories = storiesOf('Tag', module);
+
+const shortText = 'Short';
+const midText = 'A bit much longer tag';
+const longText = 'This is a looooooooooooooooooooooooooong text in order to test the tag';
```
Comment (diegolmello) [inline storybook/stories/Tag.js:12]: Use longText util

File: storybook/stories/Tag.js [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+import React from 'react';
+import { View } from 'react-native';
+import { storiesOf } from '@storybook/react-native';
+
+import Tag from '../../app/presentation/RoomItem/Tag';
+
+const stories = storiesOf('Tag', module);
+
+const shortText = 'Short';
+const midText = 'A bit much longer tag';
```
Comment (diegolmello) [inline storybook/stories/Tag.js:11]: Not needed

File: app/views/TeamChannelsView.js [hunk 26]
```diff
@@ -287,6 +290,107 @@ class TeamChannelsView extends React.Component {
 		}
 	}, 1000, true);
 
+	options = item => ([
+		{
+			title: I18n.t('Auto-join'),
+			icon: item.t === 'p' ? 'channel-private' : 'channel-public',
+			onPress: () => this.autoJoin(item),
+			right: () => (item.teamDefault ? <CustomIcon name='check' size={20} /> : null)
```
Comment (diegolmello) [inline app/views/TeamChannelsView.js:298]: Use the same icon from Messagebox thread

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/lib/database/model/migrations.js [hunk 27]
```diff
@@ -224,6 +224,17 @@ export default schemaMigrations({
 					]
 				})
 			]
+		},
+		{
+			toVersion: 14,
```
Comment (diegolmello) [inline app/lib/database/model/migrations.js:229]: Remove it

**Review by diegolmello — APPROVED** [review]
